# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
original_data=pd.read_csv('D:/Vortex Mods/homework/3.清洗数据/e_commerce.csv')

## 评估数据

In [2]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [3]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

### 首先观察数据，发现InvoiceData应该改为日期型 和 CustomerID需要改为字符型， 发现Description和CustomerID列存在缺失值。

In [4]:
original_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

#### 再观察数据结构符合三大规则。

### 观察缺失值

#### 先观察Description中的缺失值。

In [5]:
original_data[original_data['Description'].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

观察Description中的缺失值。发现当Description为缺失值时UnitPrice在结果中都为0，检验推论是否成立。并且探究Description与CustomerID是否同时全为缺失值

In [6]:
original_data[(original_data['Description'].isnull()) & (original_data['CustomerID'].isnull())]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

In [7]:
original_data[(original_data['Description'].isnull()) & (original_data['UnitPrice'] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

经检验发现推论成立发现当Description为缺失值时UnitPrice在结果中都为0，Description与CustomerID同时全为缺失值。

因为分析的目的是，根据市场销售数据，挖掘畅销产品，而产品名属于重要数据，所以Description缺失值需要后续删除。

In [8]:
original_data[original_data['CustomerID'].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

### 分析数据是否存在错误

In [9]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

发现Quantity和UnitPrice存在负数数据，分析存在负数数据的原因

In [10]:
original_data[original_data['Quantity'] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

检验，观察发现当Quantity为负数时InvoiceNo首字母都为C

In [11]:
original_data[(original_data['Quantity'] < 0) & (original_data['InvoiceNo'].str[0] == 'C')]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[9288 rows x 8 columns]

In [12]:
original_data[(original_data['Quantity'] < 0) & (original_data['InvoiceNo'].str[0] != 'C')]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

In [13]:
original_data[(original_data['Quantity'] < 0) & (original_data['InvoiceNo'].str[0] != 'C') & (original_data['UnitPrice'] == 0)]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

In [14]:
original_data[original_data['UnitPrice'] < 0]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

检验发现当Quantity为负数时InvoiceNo首字母不全为C但是不为C的项UnitPrice都为零，由于UnitPrice，全为零，但是Quantity不为零，可能是错误数据，需要删除。
其他InvoiceNo首字母为C的交易且UnitPrice不为零的值需要加到分析中，这一部分是由于退回的商品造成的销售量减小需要计算。   
当UnitPrice小于0时，表示Adjust bad debt，需要删除该数据。

#### 检查不一致数据

检查国家名称是否存在不同名称表示相同意义的情况,使用Series中的 .values_counts()方法

In [15]:
original_data['Country'].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


发现(United Kingdom,UK,U.K.)  (USA,United States)表示相同含义需要修改意义

## 清理数据

- 修改InvoiceData应该改为日期型 和 CustomerID需要改字符型
- Description缺失值需要后续删除
- 由于UnitPrice，全为零，但是Quantity不为零，可能是错误数据，需要删除
- UnitPrice小于0时，表示Adjust bad debt，需要删除该数据
- 发现(United Kingdom,UK,U.K.) (USA,United States)表示相同含义需要修改意义


In [16]:
clean_data = original_data.copy() 

使用pd.to_datetime('列')方法修改格式

In [17]:
clean_data['InvoiceDate'] = pd.to_datetime(clean_data['InvoiceDate'])

In [18]:
clean_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [19]:
clean_data['CustomerID'] = clean_data['CustomerID'].astype(str)
clean_data['CustomerID']

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

发现字符型字符串有'.0'的数值需要删除  
Series中有slice方法可以使我们保留字符串的一部分

In [20]:
clean_data['CustomerID'] = clean_data['CustomerID'].str.slice(0,-2)  #和切片一样，第一个数值表示起始值，包含在内，第二个数值表示结束值，不包含在内
clean_data['CustomerID']

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

接下来要删除Description缺失值  使用dropna方法

In [21]:
clean_data.dropna(subset=['Description'],inplace=True)

In [22]:
clean_data[original_data['Description'].isnull()]

C:\Users\22247\AppData\Local\Temp\ipykernel_19136\2957258250.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  clean_data[original_data['Description'].isnull()]


Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

删除当UnitPrice，全为零，但是Quantity不为零，可能是错误数据

In [23]:
clean_data[(clean_data['UnitPrice'] == 0) & (clean_data['Quantity'] != 0)]

InvoiceNo StockCode                   Description  Quantity  \
6391      536941     22734                        amazon        20   
6392      536942     22139                        amazon        15   
7313      537032     21275                             ?       -30   
9302      537197     22841  ROUND CAKE TIN VINTAGE GREEN         1   
13217     537425    84968F                         check       -20   
...          ...       ...                           ...       ...   
535336    581213     22576                         check       -30   
536908    581226     23090                       missing      -338   
538504    581406    46000M  POLYESTER FILLER PAD 45x45cm       240   
538505    581406    46000S  POLYESTER FILLER PAD 40x40cm       300   
538919    581422     23169                       smashed      -235   

               InvoiceDate  UnitPrice CustomerID         Country  
6391   2010-12-03 12:08:00        0.0          n  United Kingdom  
6392   2010-12-03 12:08:00        0.0          n  United Kingdom  
7313   2010-12-03 16:50:00        0.0          n  United Kingdom  
9302   2010-12-05 14:02:00        0.0      12647         Germany  
13217  2010-12-06 15:35:00        0.0          n  United Kingdom  
...                    ...        ...        ...             ...  
535336 2011-12-07 18:38:00        0.0          n  United Kingdom  
536908 2011-12-08 09:56:00        0.0          n  United Kingdom  
538504 2011-12-08 13:58:00        0.0          n  United Kingdom  
538505 2011-12-08 13:58:00        0.0          n  United Kingdom  
538919 2011-12-08 15:24:00        0.0          n  United Kingdom  

[1061 rows x 8 columns]

In [24]:
clean_data = clean_data[clean_data['UnitPrice'] != 0]

In [25]:
clean_data[(clean_data['UnitPrice'] == 0) & (clean_data['Quantity'] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

UnitPrice小于0时，表示Adjust bad debt，需要删除该数据

In [26]:
clean_data = clean_data[clean_data['UnitPrice'] >= 0]

In [27]:
clean_data[clean_data['UnitPrice'] < 0]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

发现(United Kingdom,UK,U.K.) (USA,United States)表示相同含义需要修改意义

In [28]:
clean_data['Country'] = clean_data['Country'].replace(['United Kingdom','UK','U.K.'],'United Kingdom')

In [29]:
clean_data['Country'] = clean_data['Country'].replace('United States','USA')

In [30]:
clean_data['Country'].value_counts()

Country
United Kingdom          492979
Germany                   9493
France                    8556
EIRE                      8192
Spain                     2532
Netherlands               2367
Belgium                   2069
Switzerland               2001
Portugal                  1519
Australia                 1256
Norway                    1085
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
China                      288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         57


In [31]:
clean_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 539392 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    539392 non-null  object        
 1   StockCode    539392 non-null  object        
 2   Description  539392 non-null  object        
 3   Quantity     539392 non-null  int64         
 4   InvoiceDate  539392 non-null  datetime64[ns]
 5   UnitPrice    539392 non-null  float64       
 6   CustomerID   539392 non-null  object        
 7   Country      539392 non-null  object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 37.0+ MB


In [32]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [33]:
clean_data.describe()

Quantity                    InvoiceDate      UnitPrice
count  539392.000000                         539392  539392.000000
mean        9.845904  2011-07-04 16:40:36.232350208       4.673648
min    -80995.000000            2010-12-01 08:26:00       0.001000
25%         1.000000            2011-03-28 11:59:00       1.250000
50%         3.000000            2011-07-20 11:50:00       2.080000
75%        10.000000            2011-10-19 11:49:00       4.130000
max     80995.000000            2011-12-09 12:50:00   38970.000000
std       215.412652                            NaN      94.614722

In [34]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [35]:
clean_data[clean_data['Quantity'] < 0].min()

InvoiceNo                             C536379
StockCode                               10133
Description     50'S CHRISTMAS GIFT BAG LARGE
Quantity                               -80995
InvoiceDate               2010-12-01 09:41:00
UnitPrice                                0.01
CustomerID                              12346
Country                             Australia
dtype: object

## 保存清理后的数据

In [36]:
#clean_data.to_csv ('e_commerce_clean.csv',index=False)

## 分析数据

现在需要根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

In [37]:
clean_data_analyse = clean_data.copy()

In [38]:
clean_data_analyse

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-12-09 12:50:00       0.85      12680          France  
541905 2011-12-09 12:50:00       2.10      12680          France  
541906 2011-12-09 12:50:00       4.15      12680          France  
541907 2011-12-09 12:50:00       4.15      12680          France  
541908 2011-12-09 12:50:00       4.95      12680          France  

[539392 rows x 8 columns]

In [39]:
clean_data_analyse.query('(StockCode =="10002")&(UnitPrice == 1.63)')

InvoiceNo StockCode                  Description  Quantity  \
63874     541592     10002  INFLATABLE POLITICAL GLOBE          1   
68824     541878     10002  INFLATABLE POLITICAL GLOBE          1   
92623     544201     10002  INFLATABLE POLITICAL GLOBE          1   
92755     544205     10002  INFLATABLE POLITICAL GLOBE          1   
106195    545316     10002  INFLATABLE POLITICAL GLOBE          2   
111324    545717     10002  INFLATABLE POLITICAL GLOBE          2   
120160    546649     10002  INFLATABLE POLITICAL GLOBE          1   

               InvoiceDate  UnitPrice CustomerID         Country  
63874  2011-01-19 15:08:00       1.63          n  United Kingdom  
68824  2011-01-24 10:28:00       1.63          n  United Kingdom  
92623  2011-02-17 10:11:00       1.63          n  United Kingdom  
92755  2011-02-17 10:31:00       1.63          n  United Kingdom  
106195 2011-03-01 14:14:00       1.63          n  United Kingdom  
111324 2011-03-07 10:15:00       1.63          n  United Kingdom  
120160 2011-03-15 14:17:00       1.63          n  United Kingdom

In [40]:
clean_data_analyse['price_sum'] = clean_data_analyse['Quantity']*clean_data_analyse['UnitPrice']
clean_data_analyse

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice CustomerID         Country  price_sum  
0      2010-12-01 08:26:00       2.55      17850  United Kingdom      15.30  
1      2010-12-01 08:26:00       3.39      17850  United Kingdom      20.34  
2      2010-12-01 08:26:00       2.75      17850  United Kingdom      22.00  
3      2010-12-01 08:26:00       3.39      17850  United Kingdom      20.34  
4      2010-12-01 08:26:00       3.39      17850  United Kingdom      20.34  
...                    ...        ...        ...             ...        ...  
541904 2011-12-09 12:50:00       0.85      12680          France      10.20  
541905 2011-12-09 12:50:00       2.10      12680          France      12.60  
541906 2011-12-09 12:50:00       4.15      12680          France      16.60  
541907 2011-12-09 12:50:00       4.15      12680          France      16.60  
541908 2011-12-09 12:50:00       4.95      12680          France      14.85  

[539392 rows x 9 columns]

In [41]:
clean_data_analyse_hot = clean_data_analyse.groupby('StockCode')[['Quantity','price_sum']].sum()
clean_data_analyse_hot

Quantity  price_sum
StockCode                        
10002              860     759.89
10080              303     119.09
10120              193      40.53
10123C               5       3.25
10124A              16       6.72
...                ...        ...
gift_0001_20        10     167.05
gift_0001_30         7     175.53
gift_0001_40         3     100.70
gift_0001_50         4     167.56
m                    1       2.55

[3938 rows x 2 columns]

In [42]:
clean_data_analyse_hot.sort_values(by=['price_sum','Quantity'],ascending=[False,False])

Quantity   price_sum
StockCode                         
DOT                705  206245.480
22423            13022  164762.190
47566            18018   98302.980
85123A           35082   97894.500
85099B           47359   92356.030
...                ...         ...
D                -1194   -5696.220
BANK CHARGES       -13   -7175.639
CRUK               -16   -7933.430
M                 3158  -68674.190
AMAZONFEE          -30 -221520.500

[3938 rows x 2 columns]

In [43]:
clean_data_analyse_hot.sort_values(by=['Quantity','price_sum'],ascending=[False,False])

Quantity  price_sum
StockCode                     
22197         56450   50987.47
84077         53847   13587.93
85099B        47359   92356.03
84879         36381   58959.73
21212         36039   21059.72
...             ...        ...
CRUK            -16   -7933.43
21645           -24     -39.60
AMAZONFEE       -30 -221520.50
S               -59   -3049.39
D             -1194   -5696.22

[3938 rows x 2 columns]

以`price_sum`为第一标准`Quantity`为第二标准，进行降序排序。再以相反的排序标准排序分析。发现`StockCode`为`DOT`的商品，以较少的销售量带来了巨大的经济效益，应当列为推销商品。`StockCode`为`22423`,`47566`,`85123A`,`85099B`等产品是热销产品，且具有相当的经济效益。
现在需要寻找两个排序方式的上四分位数之上的所有商品。

In [44]:
clean_data_analyse_hot = clean_data_analyse_hot.sort_values(by=['price_sum','Quantity'],ascending=[False,False])
clean_data_analyse_hot = clean_data_analyse_hot.reset_index()
clean_data_analyse_hot

StockCode  Quantity   price_sum
0              DOT       705  206245.480
1            22423     13022  164762.190
2            47566     18018   98302.980
3           85123A     35082   97894.500
4           85099B     47359   92356.030
...            ...       ...         ...
3933             D     -1194   -5696.220
3934  BANK CHARGES       -13   -7175.639
3935          CRUK       -16   -7933.430
3936             M      3158  -68674.190
3937     AMAZONFEE       -30 -221520.500

[3938 rows x 3 columns]

In [45]:
clean_data_analyse_hot[['Quantity','price_sum']].quantile(0.9)

Quantity     3394.400
price_sum    6405.027
Name: 0.9, dtype: float64

In [46]:
clean_data_analyse_hot_price = clean_data_analyse_hot.head(300)
clean_data_analyse_hot_price

StockCode  Quantity  price_sum
0         DOT       705  206245.48
1       22423     13022  164762.19
2       47566     18018   98302.98
3      85123A     35082   97894.50
4      85099B     47359   92356.03
..        ...       ...        ...
295     22940      1896    7854.50
296     22667      2808    7840.40
297     22352      3077    7832.92
298     21136      4700    7815.11
299     23238      1846    7759.96

[300 rows x 3 columns]

In [47]:
clean_data_analyse_hot = clean_data_analyse_hot.sort_values(by=['Quantity','price_sum'],ascending=[False,False])
clean_data_analyse_hot = clean_data_analyse_hot.reset_index()
clean_data_analyse_hot

index  StockCode  Quantity  price_sum
0        11      22197     56450   50987.47
1       143      84077     53847   13587.93
2         4     85099B     47359   92356.03
3         8      84879     36381   58959.73
4        82      21212     36039   21059.72
...     ...        ...       ...        ...
3933   3935       CRUK       -16   -7933.43
3934   3928      21645       -24     -39.60
3935   3937  AMAZONFEE       -30 -221520.50
3936   3932          S       -59   -3049.39
3937   3933          D     -1194   -5696.22

[3938 rows x 4 columns]

In [48]:
clean_data_analyse_hot_quantity = clean_data_analyse_hot.head(300)
clean_data_analyse_hot_quantity

index StockCode  Quantity  price_sum
0       11     22197     56450   50987.47
1      143     84077     53847   13587.93
2        4    85099B     47359   92356.03
3        8     84879     36381   58959.73
4       82     21212     36039   21059.72
..     ...       ...       ...        ...
295    187     22041      4336   11260.81
296    241     84375      4308    9376.85
297    410     21670      4301    6045.04
298    370     84988      4299    6653.85
299    421     21673      4292    5927.80

[300 rows x 4 columns]

现在合并'clean_data_analyse_hot_quantity'和'clean_data_analyse_hot_price'两个数据集删除重复值即可得出热销商品。

In [49]:
#使用concat方法纵向拼接两个个数据即可。
hot_goods = pd.concat([clean_data_analyse_hot_quantity,clean_data_analyse_hot_price])
hot_goods

index StockCode  Quantity  price_sum
0     11.0     22197     56450   50987.47
1    143.0     84077     53847   13587.93
2      4.0    85099B     47359   92356.03
3      8.0     84879     36381   58959.73
4     82.0     21212     36039   21059.72
..     ...       ...       ...        ...
295    NaN     22940      1896    7854.50
296    NaN     22667      2808    7840.40
297    NaN     22352      3077    7832.92
298    NaN     21136      4700    7815.11
299    NaN     23238      1846    7759.96

[600 rows x 4 columns]

In [50]:
hot_goods = hot_goods.drop_duplicates(subset='StockCode')
hot_goods

index StockCode  Quantity  price_sum
0     11.0     22197     56450   50987.47
1    143.0     84077     53847   13587.93
2      4.0    85099B     47359   92356.03
3      8.0     84879     36381   58959.73
4     82.0     21212     36039   21059.72
..     ...       ...       ...        ...
293    NaN     21770      1635    7909.08
295    NaN     22940      1896    7854.50
296    NaN     22667      2808    7840.40
297    NaN     22352      3077    7832.92
299    NaN     23238      1846    7759.96

[427 rows x 4 columns]

In [51]:
#hot_goods = hot_goods.drop('index',axis=1)
hot_goods

index StockCode  Quantity  price_sum
0     11.0     22197     56450   50987.47
1    143.0     84077     53847   13587.93
2      4.0    85099B     47359   92356.03
3      8.0     84879     36381   58959.73
4     82.0     21212     36039   21059.72
..     ...       ...       ...        ...
293    NaN     21770      1635    7909.08
295    NaN     22940      1896    7854.50
296    NaN     22667      2808    7840.40
297    NaN     22352      3077    7832.92
299    NaN     23238      1846    7759.96

[427 rows x 4 columns]

In [52]:
hot_goods = hot_goods.drop('index',axis=1)
hot_goods

StockCode  Quantity  price_sum
0       22197     56450   50987.47
1       84077     53847   13587.93
2      85099B     47359   92356.03
3       84879     36381   58959.73
4       21212     36039   21059.72
..        ...       ...        ...
293     21770      1635    7909.08
295     22940      1896    7854.50
296     22667      2808    7840.40
297     22352      3077    7832.92
299     23238      1846    7759.96

[427 rows x 3 columns]

In [53]:
hot_goods = hot_goods.reset_index()

In [54]:
hot_goods

index StockCode  Quantity  price_sum
0        0     22197     56450   50987.47
1        1     84077     53847   13587.93
2        2    85099B     47359   92356.03
3        3     84879     36381   58959.73
4        4     21212     36039   21059.72
..     ...       ...       ...        ...
422    293     21770      1635    7909.08
423    295     22940      1896    7854.50
424    296     22667      2808    7840.40
425    297     22352      3077    7832.92
426    299     23238      1846    7759.96

[427 rows x 4 columns]

In [55]:
#观数据是否存在销售量极少但是利润极高的商品，我们认为这种商品的销售策略是需要调整的。
hot_goods.query('Quantity <= 50')

index StockCode  Quantity  price_sum
367    193         B         1   11062.06

In [58]:
#为数据分箱，间隔为150
sale_volume=list(range(0,57000,300))
sale_volume

[0,
 300,
 600,
 900,
 1200,
 1500,
 1800,
 2100,
 2400,
 2700,
 3000,
 3300,
 3600,
 3900,
 4200,
 4500,
 4800,
 5100,
 5400,
 5700,
 6000,
 6300,
 6600,
 6900,
 7200,
 7500,
 7800,
 8100,
 8400,
 8700,
 9000,
 9300,
 9600,
 9900,
 10200,
 10500,
 10800,
 11100,
 11400,
 11700,
 12000,
 12300,
 12600,
 12900,
 13200,
 13500,
 13800,
 14100,
 14400,
 14700,
 15000,
 15300,
 15600,
 15900,
 16200,
 16500,
 16800,
 17100,
 17400,
 17700,
 18000,
 18300,
 18600,
 18900,
 19200,
 19500,
 19800,
 20100,
 20400,
 20700,
 21000,
 21300,
 21600,
 21900,
 22200,
 22500,
 22800,
 23100,
 23400,
 23700,
 24000,
 24300,
 24600,
 24900,
 25200,
 25500,
 25800,
 26100,
 26400,
 26700,
 27000,
 27300,
 27600,
 27900,
 28200,
 28500,
 28800,
 29100,
 29400,
 29700,
 30000,
 30300,
 30600,
 30900,
 31200,
 31500,
 31800,
 32100,
 32400,
 32700,
 33000,
 33300,
 33600,
 33900,
 34200,
 34500,
 34800,
 35100,
 35400,
 35700,
 36000,
 36300,
 36600,
 36900,
 37200,
 37500,
 37800,
 38100,
 38400,
 38700,


In [62]:
def Sales_heat(x):
    if x <= 15000:
        return '销量一般'
    elif x <= 30000:
        return '销量较好'
    elif x <=57000:
        return '销量极好'
pd1=pd.Series(sale_volume)
pd1.apply(Sales_heat)
    

0      销量一般
1      销量一般
2      销量一般
3      销量一般
4      销量一般
       ... 
185    销量极好
186    销量极好
187    销量极好
188    销量极好
189    销量极好
Length: 190, dtype: object

In [ ]:
#输出数据
hot_goods.to_csv('评估和清理英国电商公司销售数据的热销商品.csv',index=False)

In [ ]:
#stockcode_price = clean_data_analyse[['StockCode','UnitPrice']]
#stockcode_price

In [ ]:
#stockcode_price = stockcode_price.drop_duplicates()

In [ ]:
#stockcode_price.value_counts()

In [ ]:
#stockcode_price['']